File: Namyata/MaxProfit.py
# You are given an array prices where prices[i] is the price of a given stock on the ith day.
#
# You want to maximize your profit by choosing a single day to buy one stock and choosing a different day in the future to sell that stock.
#
# Return the maximum profit you can achieve from this transaction. If you cannot achieve any profit, return 0.



import unittest

def maxProfit(prices):
    left = 0  # Buy
    right = 1  # Sell
    max_profit = 0
    while right < len(prices):
        current_profit = prices[right] - prices[left]  # our current Profit
        if prices[left] < prices[right]:
            max_profit = max(current_profit, max_profit)
        else:
            left = right
        right += 1
    return max_profit

class MaxProfitTestCase(unittest.TestCase):
    def test_profit_possible(self):
        prices = [7, 1, 5, 3, 6, 4]
        expected_output = 5
        self.assertEqual(maxProfit(prices), expected_output)

    def test_no_profit_possible(self):
        prices = [7, 6, 4, 3, 1]
        expected_output = 0
        self.assertEqual(maxProfit(prices), expected_output)

    def test_empty_array(self):
        prices = []
        expected_output = 0
        self.assertEqual(maxProfit(prices), expected_output)

    def test_single_element_array(self):
        prices = [5]
        expected_output = 0
        self.assertEqual(maxProfit(prices), expected_output)

    def test_constant_price(self):
        prices = [3, 3, 3, 3, 3]
        expected_output = 0
        self.assertEqual(maxProfit(prices), expected_output)

if __name__ == '__main__':
    unittest.main()

File: CTests/max_profit.c
#include <stdio.h>
#include <stdbool.h>
#include <stdlib.h>

int maxProfit(int* prices, int pricesSize) {
    int left = 0;  // Buy
    int right = 1; // Sell
    int maxProfit = 0;

    while (right < pricesSize) {
        int currentProfit = prices[right] - prices[left]; // Current profit

        if (prices[left] < prices[right]) {
            maxProfit = (currentProfit > maxProfit) ? currentProfit : maxProfit;
        } else {
            left = right;
        }

        right++;
    }

    return maxProfit;
}

bool testProfitPossible() {
    int prices[] = {7, 1, 5, 3, 6, 4};
    int pricesSize = sizeof(prices) / sizeof(prices[0]);
    int expectedOutput = 5;
    int result = maxProfit(prices, pricesSize);
    if (result == expectedOutput) {
        printf("Profit Possible Case: Passed\n");
        return true;
    } else {
        printf("Profit Possible Case: Failed\n");
        return false;
    }
}

bool testNoProfitPossible() {
    int prices[] = {7, 6, 4, 3, 1};
    int pricesSize = sizeof(prices) / sizeof(prices[0]);
    int expectedOutput = 0;
    int result = maxProfit(prices, pricesSize);
    if (result == expectedOutput) {
        printf("No Profit Possible Case: Passed\n");
        return true;
    } else {
        printf("No Profit Possible Case: Failed\n");
        return false;
    }
}

bool testEmptyArray() {
    int prices[] = {};
    int pricesSize = 0;
    int expectedOutput = 0;
    int result = maxProfit(prices, pricesSize);
    if (result == expectedOutput) {
        printf("Empty Array Case: Passed\n");
        return true;
    } else {
        printf("Empty Array Case: Failed\n");
        return false;
    }
}

bool testSingleElementArray() {
    int prices[] = {5};
    int pricesSize = sizeof(prices) / sizeof(prices[0]);
    int expectedOutput = 0;
    int result = maxProfit(prices, pricesSize);
    if (result == expectedOutput) {
        printf("Single Element Array Case: Passed\n");
        return true;
    } else {
        printf("Single Element Array Case: Failed\n");
        return false;
    }
}

bool testConstantPrice() {
    int prices[] = {3, 3, 3, 3, 3};
    int pricesSize = sizeof(prices) / sizeof(prices[0]);
    int expectedOutput = 0;
    int result = maxProfit(prices, pricesSize);
    if (result == expectedOutput) {
        printf("Constant Price Case: Passed\n");
        return true;
    } else {
        printf("Constant Price Case: Failed\n");
        return false;
    }
}


int main() {
    bool success = true;
    success &= testProfitPossible();
    success &= testNoProfitPossible();
    success &= testEmptyArray();
    success &= testSingleElementArray();
    success &= testConstantPrice();

    if (success) {
        printf("All test cases passed!\n");
    } else {
        printf("Some test cases failed!\n");
    }

    return 0;
}
